File: backend/internal/anki/queries.go
package anki

import (
	"database/sql"
	"fmt"
	"strings"
)

// fieldSeparator is the ASCII "unit separator" byte Anki uses to join a note's field
// values into the single "flds" column.
const fieldSeparator = "\x1f"

// readNotes reads the notes table. Each note's fields arrive as one separator-joined
// string; we split it and pair each value with its field name from the note's model
// (falling back to "Field N" when the model has fewer fields than the stored note).
// It returns the notes slice plus a by-ID lookup used later when building cards.
func readNotes(db *sql.DB, noteTypes []NoteType) ([]Note, map[string]Note, error) {
	modelByID := map[string]NoteType{}
	for _, model := range noteTypes {
		modelByID[model.ID] = model
	}
	rows, err := db.Query(`SELECT id, guid, mid, mod, usn, tags, flds, sfld FROM notes`)
	if err != nil {
		return nil, nil, err
	}
	defer rows.Close()

	notes := []Note{}
	byID := map[string]Note{}
	for rows.Next() {
		var id, guid, mid, tags, flds, sfld string
		var mod, usn int64
		if err := rows.Scan(&id, &guid, &mid, &mod, &usn, &tags, &flds, &sfld); err != nil {
			return nil, nil, err
		}
		rawFields := strings.Split(flds, fieldSeparator)
		model := modelByID[mid]
		fieldOrder := make([]string, 0, len(rawFields))
		fields := map[string]string{}
		for i, value := range rawFields {
			name := fmt.Sprintf("Field %d", i+1)
			if i < len(model.Fields) {
				name = model.Fields[i].Name
			}
			fieldOrder = append(fieldOrder, name)
			fields[name] = value
		}
		note := Note{
			ID:         id,
			GUID:       guid,
			NoteTypeID: mid,
			SortField:  sfld,
			Tags:       splitTags(tags),
			Fields:     fields,
			FieldOrder: fieldOrder,
			RawFields:  rawFields,
			Mod:        mod,
			USN:        usn,
		}
		notes = append(notes, note)
		byID[id] = note
	}
	return notes, byID, rows.Err()
}

// readCards reads the cards table and, for each card, renders a plain-text front/back
// preview from its note and template (see previewFrontBack in template.go).
func readCards(db *sql.DB, decks []Deck, noteTypes []NoteType, notes map[string]Note) ([]Card, error) {
	modelByID := map[string]NoteType{}
	for _, model := range noteTypes {
		modelByID[model.ID] = model
	}
	rows, err := db.Query(`SELECT id, nid, did, ord, mod, usn, type, queue, due, ivl, factor, reps, lapses, left, odid, flags, data FROM cards`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	cards := []Card{}
	for rows.Next() {
		var id, nid, did, data string
		var ord, mod, usn, typ, queue, due, ivl, factor, reps, lapses, left, odid, flags int64
		if err := rows.Scan(&id, &nid, &did, &ord, &mod, &usn, &typ, &queue, &due, &ivl, &factor, &reps, &lapses, &left, &odid, &flags, &data); err != nil {
			return nil, err
		}
		note := notes[nid]
		model := modelByID[note.NoteTypeID]
		templateName := ""
		if ord >= 0 && int(ord) < len(model.Templates) {
			templateName = model.Templates[ord].Name
		}
		front, back := previewFrontBack(note, model, ord)
		cards = append(cards, Card{
			ID:             id,
			NoteID:         nid,
			DeckID:         did,
			Ord:            ord,
			Type:           typ,
			Queue:          queue,
			Due:            due,
			Interval:       ivl,
			Factor:         factor,
			Reps:           reps,
			Lapses:         lapses,
			Left:           left,
			OriginalDeckID: zeroEmpty(odid),
			Flags:          flags,
			Data:           data,
			TemplateName:   templateName,
			Front:          front,
			Back:           back,
			Raw: map[string]any{
				"mod": mod,
				"usn": usn,
			},
		})
	}
	return cards, rows.Err()
}

// readReviewLogs reads the revlog table (the per-review history). It is best-effort:
// callers treat an error here as a warning rather than a failed import.
func readReviewLogs(db *sql.DB) ([]ReviewLog, error) {
	rows, err := db.Query(`SELECT id, cid, usn, ease, ivl, lastIvl, factor, time, type FROM revlog`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []ReviewLog{}
	for rows.Next() {
		var id, cid string
		var usn, ease, ivl, lastIvl, factor, reviewTime, typ int64
		if err := rows.Scan(&id, &cid, &usn, &ease, &ivl, &lastIvl, &factor, &reviewTime, &typ); err != nil {
			return nil, err
		}
		out = append(out, ReviewLog{ID: id, CardID: cid, USN: usn, Ease: ease, Interval: ivl, LastIvl: lastIvl, Factor: factor, Time: reviewTime, Type: typ})
	}
	return out, rows.Err()
}

File: src/utils/apkg/queries.ts
// notes/cards/revlog table readers, mirroring backend/internal/anki/queries.go. Row order is
// whatever SQLite returns for an unordered SELECT — both implementations scan the same file,
// so they see the same order.

import type { Database } from "sql.js";
import type { ApkgCard, ApkgNote, ApkgNoteType, ApkgReviewLog } from "./types";
import { omitEmpty, omitZero, splitTags, zeroEmpty } from "./coerce";
import { previewFrontBack } from "./template";

/** fieldSeparator is the ASCII "unit separator" Anki joins a note's field values with. */
const fieldSeparator = "\x1f";

/** str/int coerce SQLite column values the way Go's Scan into string/int64 does. */
function str(v: unknown): string {
  return v === null || v === undefined ? "" : String(v);
}

function int(v: unknown): number {
  return typeof v === "number" ? v : v === null || v === undefined ? 0 : Number(v);
}

/** eachRow runs stmt to completion, freeing it even when a row callback throws. */
function eachRow(db: Database, sql: string, onRow: (row: unknown[]) => void): void {
  const stmt = db.prepare(sql);
  try {
    while (stmt.step()) onRow(stmt.get());
  } finally {
    stmt.free();
  }
}

/**
 * readNotes reads the notes table, splitting each note's separator-joined field string and
 * pairing values with field names from its model ("Field N" when the model is unknown or has
 * fewer fields). Returns the notes plus a by-id lookup used when building cards.
 */
export function readNotes(db: Database, noteTypes: ApkgNoteType[]): [ApkgNote[], Map<string, ApkgNote>] {
  const modelById = new Map(noteTypes.map((m) => [m.id, m]));
  const notes: ApkgNote[] = [];
  const byId = new Map<string, ApkgNote>();
  eachRow(db, "SELECT id, guid, mid, mod, usn, tags, flds, sfld FROM notes", (row) => {
    const [id, guid, mid, mod, usn, tags, flds, sfld] = row;
    const rawFields = str(flds).split(fieldSeparator);
    const model = modelById.get(str(mid));
    const fieldOrder: string[] = [];
    const fields: Record<string, string> = {};
    rawFields.forEach((value, i) => {
      const name = model && i < model.fields.length ? model.fields[i].name : `Field ${i + 1}`;
      fieldOrder.push(name);
      fields[name] = value;
    });
    const note: ApkgNote = {
      id: str(id),
      guid: str(guid),
      noteTypeId: str(mid),
      sortField: omitEmpty(str(sfld)),
      tags: splitTags(str(tags)),
      fields,
      fieldOrder,
      rawFields,
      mod: omitZero(int(mod)),
      usn: omitZero(int(usn)),
    };
    notes.push(note);
    byId.set(note.id, note);
  });
  return [notes, byId];
}

/** readCards reads the cards table, rendering a plain-text front/back preview per card. */
export function readCards(db: Database, noteTypes: ApkgNoteType[], notes: Map<string, ApkgNote>): ApkgCard[] {
  const modelById = new Map(noteTypes.map((m) => [m.id, m]));
  const cards: ApkgCard[] = [];
  eachRow(
    db,
    "SELECT id, nid, did, ord, mod, usn, type, queue, due, ivl, factor, reps, lapses, left, odid, flags, data FROM cards",
    (row) => {
      const [id, nid, did, ord, mod, usn, typ, queue, due, ivl, factor, reps, lapses, left, odid, flags, data] = row;
      const note = notes.get(str(nid)) ?? {
        id: "", guid: "", noteTypeId: "", tags: [], fields: {}, fieldOrder: [], rawFields: [],
      };
      const model = modelById.get(note.noteTypeId);
      const ordNum = int(ord);
      const templates = model?.templates ?? [];
      const templateName = ordNum >= 0 && ordNum < templates.length ? templates[ordNum].name : "";
      const [front, back] = previewFrontBack(note, model, ordNum);
      cards.push({
        id: str(id),
        noteId: str(nid),
        deckId: str(did),
        ord: ordNum,
        type: int(typ),
        queue: int(queue),
        due: int(due),
        interval: int(ivl),
        factor: int(factor),
        reps: int(reps),
        lapses: int(lapses),
        left: omitZero(int(left)),
        originalDeckId: omitEmpty(zeroEmpty(int(odid))),
        flags: omitZero(int(flags)),
        data: omitEmpty(str(data)),
        templateName: omitEmpty(templateName),
        front,
        back,
        raw: { mod: int(mod), usn: int(usn) },
      });
    }
  );
  return cards;
}

/** readReviewLogs reads the revlog table; callers treat a throw as a warning (Go parity). */
export function readReviewLogs(db: Database): ApkgReviewLog[] {
  const out: ApkgReviewLog[] = [];
  eachRow(db, "SELECT id, cid, usn, ease, ivl, lastIvl, factor, time, type FROM revlog", (row) => {
    const [id, cid, usn, ease, ivl, lastIvl, factor, time, typ] = row;
    out.push({
      id: str(id),
      cardId: str(cid),
      usn: int(usn),
      ease: int(ease),
      interval: int(ivl),
      lastInterval: int(lastIvl),
      factor: int(factor),
      time: int(time),
      type: int(typ),
    });
  });
  return out;
}
